lib/spack/spack/parser.py

```diff
@@ -0,0 +1,458 @@
+
+
+class TokenContext:
+    """Token context passed around by parsers"""
+
+    __slots__ = "token_stream", "current_token", "next_token"
+
+    def __init__(self, token_stream: Iterator[Token]):
+        self.token_stream = token_stream
+        self.current_token = None
+        self.next_token = None
+        self.advance()
+
+    def advance(self):
+        """Advance one token"""
+        self.current_token, self.next_token = self.next_token, next(self.token_stream, None)
+
+    def accept(self, kind: TokenKind):
+        """If the next token is of the specified kind, advance the stream and return True.
+        Otherwise return False.
+        """
+        if self.next_token and self.next_token.kind == kind:
+            self.advance()
+            return True
+        return False
+
+
+class SpecParser:
+    """Parse text into specs"""
+
+    __slots__ = "literal_str", "ctx"
+
+    def __init__(self, literal_str: str):
+        self.literal_str = literal_str
+        self.ctx = TokenContext(filter(lambda x: x.kind != TokenKind.WS, tokenize(literal_str)))
+
+    def tokens(self) -> List[Token]:
+        """Return the entire list of token from the initial text. White spaces are
+        filtered out.
+        """
+        return list(filter(lambda x: x.kind != TokenKind.WS, tokenize(self.literal_str)))
+
+    def next_spec(self, spec_buffer: Optional[spack.spec.Spec] = None) -> spack.spec.Spec:
+        """Return the next spec parsed from text.
+
+        Args:
+            spec_buffer: buffer where to parse the spec. If None a new one
+                will be created.
+
+        Return
+            The spec that was parsed
+        """
+        spec_buffer = spec_buffer or spack.spec.Spec()
+        root_spec = SpecNodeParser(self.ctx).parse(spec_buffer)
+        while True:
+            if self.ctx.accept(TokenKind.DEPENDENCY):
+                dependency = SpecNodeParser(self.ctx).parse(spack.spec.Spec())
+
+                if dependency == spack.spec.Spec():
+                    msg = (
+                        "this dependency sigil needs to be followed by a package name "
+                        "or a node attribute (version, variant, etc.)"
+                    )
+                    raise SpecParsingError(msg, self.ctx.current_token, self.literal_str)
+
+                root_spec._add_dependency(dependency, ())
+
+            else:
+                break
+
+        return root_spec
+
+    def all_specs(self) -> List[spack.spec.Spec]:
+        """Return all the specs that remain to be parsed"""
+        return list(iter(self.next_spec, spack.spec.Spec()))
+
+
+class SpecNodeParser:
+    """Parse a single spec node from a stream of tokens"""
+
+    __slots__ = "ctx", "has_compiler", "has_version"
+
+    def __init__(self, ctx):
+        self.ctx = ctx
+        self.has_compiler = False
+        self.has_version = False
+
+    def parse(self, spec_buffer: spack.spec.Spec) -> spack.spec.Spec:
```

Comment:
Can we call this something other than "buffer"?  A buffer always means raw bytes or chars to me -- and this thing is an object under construction.  Honestly just `spec` seems clear to me (esp. since the caller won't usually pass it).

Reply:
I'll go with `initial_spec` here. Let me know if that is fine.